### PR TITLE
[mlir][vector] Disable `vector.matrix_multiply` for scalable vectors

### PR DESCRIPTION
**Pull request by banach-space**
Disables `vector.matrix_multiply` for scalable vectors. As per the docs:

>  This is the counterpart of llvm.matrix.multiply in MLIR

I'm not aware of any use of matrix-multiply intrinsics in the context of
scalable vectors, hence disabling.

### CODE REVIEW

File: mlir/test/Dialect/Vector/invalid.mlir [hunk 1]
```diff
@@ -1862,3 +1862,16 @@ func.func @invalid_step_2d() {
   vector.step : vector<2x4xf32>
   return
 }
+
+// -----
+
+func.func @matrix_matmul_scalable(%a: vector<[4]xf64>, %b: vector<4xf64>) {
```
Comment (c-rhodes) [inline mlir/test/Dialect/Vector/invalid.mlir:1868]:
```suggestion
func.func @matrix_multiply_scalable(%a: vector<[4]xf64>, %b: vector<4xf64>) {
```

File: mlir/include/mlir/IR/CommonTypeConstraints.td [hunk 2]
```diff
@@ -494,6 +494,14 @@ class VectorOfRankAndType<list<int> allowedRanks,
   VectorOf<allowedTypes>.summary # VectorOfRank<allowedRanks>.summary,
   "::mlir::VectorType">;
 
+// Fixed-width vector where the rank is from the given `allowedRanks` list and
+// the type is from the given `allowedTypes` list
+class FixedVectorOfRankAndType<list<int> allowedRanks,
+                          list<Type> allowedTypes> : AllOfType<
```
Comment (c-rhodes) [inline mlir/include/mlir/IR/CommonTypeConstraints.td:500]: nit: indent to align
```suggestion
class FixedVectorOfRankAndType<list<int> allowedRanks,
                               list<Type> allowedTypes> : AllOfType<
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 3]
```diff
@@ -2707,7 +2707,10 @@ def Vector_MatmulOp : Vector_Op<"matrix_multiply", [Pure,
     <rhs_columns> and multiplies them. The result matrix is returned embedded in
     the result vector.
 
-    Also see:
+    Note, the semantics of the corresponding LLVM intrinsic,
+    `@llvm.matrix.multiply.*`, are not clear in the context of scalable
+    vectors. Hence, this Op is only available for fixed-width vectors. Also
+    see:
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:2713]: I just had a quick look at https://llvm.org/docs/LangRef.html#matrix-intrinsics and noticed
> These intrinsics require matrix dimensions to be passed as immediate arguments

So the shape can't represent scalable quantities. I think we can conclude these intrinsics don't support scalable vectors?

**Review by c-rhodes — COMMENTED** [review]
Thanks for addressing this, left a few minor comments, also the commit header needs fixing `s/nlir/mlir`

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td (hunk 3 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: Yup,  thanks for checking 👍🏻

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

**Pull request merged**